service/worker/workerdeployment/workflow.go

```diff
@@ -477,7 +484,26 @@ func (d *WorkflowRunner) validateAddVersionToWorkerDeployment(version string) er
 			return temporal.NewApplicationError("deployment version already registered", errVersionAlreadyExistsType)
 		}
 	}
+	return nil
+}
 
+func (d *WorkflowRunner) checkMaxVersions(ctx workflow.Context) error {
+	getMaxVersionsInDeployment := func(ctx workflow.Context) interface{} {
+		return dynamicconfig.MatchingMaxVersionsInDeployment.Get(d.dc)(d.a.namespace.Name().String())
+	}
+	intEq := func(a, b interface{}) bool {
+		aInt, _ := a.(int)
+		bInt, _ := b.(int)
+		return aInt == bInt
+	}
+	var maxVersions int
+	if err := workflow.MutableSideEffect(ctx, "getMaxVersions", getMaxVersionsInDeployment, intEq).Get(&maxVersions); err != nil {
+		d.logger.Error("error decoding max versions: ", err)
+		return err
+	}
+	if len(d.State.Versions) >= maxVersions {
```

Comment:
mixing the sideeffect and logic in one function feels a little weird, can you have one function to handle all the side effect business, returning a struct of concrete values (or a single one) and another function for the workflow logic (comparison)?